source/dmocks/repository.d

```diff
@@ -183,16 +184,67 @@ class MockRepository
     string buildExpectationError(bool checkUnmatchedExpectations = true, bool checkUnexpectedCalls = true)
     {
         auto expectationError = appender!string;
+        auto unexpectedCalls = _unexpectedCalls;
 
+        // return false to abort
+        void walkExpectations(ref Expectation expectation)
+        {
+            if (auto groupExpectation = cast(GroupExpectation) expectation)
+            {
+                auto expectations = groupExpectation.expectations.dup;
+
+                foreach (ref subExpectation; expectations)
+                {
+                    walkExpectations(subExpectation);
+                }
+                expectations = expectations.remove!(a => a is null);
+
+                if (expectations.empty)
+                {
+                    expectation = null;
+                    return;
+                }
+                auto newExpectation = new GroupExpectation;
+
+                newExpectation.expectations = expectations;
+                newExpectation.ordered = groupExpectation.ordered;
+                newExpectation.repeatInterval = groupExpectation.repeatInterval;
+                expectation = newExpectation;
+                return;
+            }
+            if (auto callExpectation = cast(CallExpectation) expectation)
+            {
+                if (callExpectation.satisfied) return;
+                foreach (i, unexpectedCall; unexpectedCalls)
+                {
+                    if (callExpectation.name.matches(unexpectedCall.name))
+                    {
+                        // name matches: assume they are meant to match up, generate parameter diff
+                        expectationError ~= "\n";
+                        expectationError ~= CallExpectationDiff(unexpectedCall, callExpectation);
+                        expectation = null;
+                        unexpectedCalls = unexpectedCalls.remove(i);
```

Comment:
Is it safe to remove elements during the iteration?

I have an alternative solution that may work, but I'm not insisting it is better, but it also eliminates the `if` in the `foreach` (it filters the elements before the loop).
```d
import std.range : dropBack;

if (callExpectation.satisfied)
{
    return;
}
alias pred = call => callExpectation.name.matches(call.name);
auto matched = unexpectedCalls.partition!(pred, SwapStrategy.stable);

foreach (unexpectedCall; unexpectedCalls.dropBack(matched.length))
{
    // name matches: assume they are meant to match up, generate parameter diff
    expectationError ~= "\n";
    expectationError ~= CallExpectationDiff(unexpectedCall, callExpectation);
    expectation = null;
}
unexpectedCalls = matched;
return;
```

Reply:
habs so ähnlich gemacht